# Python Programming - Objects, Variables & Data Type

## Jupyter Notebook
Jupyter Notebook is an open-source web application that allows users to create and share interactive documents, called "notebooks," containing live code, visualizations, explanatory text, and multimedia elements. It is widely used for data analysis, data visualization, machine learning, scientific computing, and interactive programming.

### User interface
The Jupyer notebook works on two modes: **edit mode** and **command mode**.

#### Edit mode
Edit mode is indicated by a green cell border and a cursor showing in the editor area:
<img src="https://jupyter-notebook.readthedocs.io/en/stable/_images/edit_mode.png" width=750>
- Edit the content of a cell
- Press <code>enter</code> (<code>return</code> for Mac OS) to enable edit mode, and <code>esc</code> to quit editing

#### Command mode
Command mode is indicated by a grey cell border with a blue left margin:
<img src="https://jupyter-notebook.readthedocs.io/en/stable/_images/command_mode.png" width=762>
- Notebook level actions: insert/remove cells, and change cell types, etc.
- Keyboard shortcuts
    - <code>A</code>: insert a cell above the current cell
    - <code>B</code>: insert a cell below the current cell
    - <code>D</code>, <code>D</code>: delete the current cell. press D twice
    - <code>M</code>: change the cell to markdown mode
    - <code>Y</code>: change the cell to code mode

#### Keyboard shortcuts
- <code>ctrl</code> + <code>enter</code> (<code>control</code> + <code>return</code> for Mac OS): run current cell
- <code>shift</code> + <code>enter</code> (<code>shift</code> + <code>return</code> for Mac OS): run current cell, and move to the cell below, if no cell below it will create new cell automatically. 

Once a cell (code) is executed, there will be a number appearing in the square ( [ ] ) bracket in the In (input) cell which denotes the order of the cell execution.

###  Detailed documentation shall be found in 
- <a href="https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/examples_index.html"><b>Jupyter Notebook Documentation</b></a> 

## Introduction to Python programming
Python is a simple language, where coding looks like speaking/writing English. It talks with words and
sentences, rather than characters (CHAR). Python shall be very easily installed on any computer.

In [ ]:
a_week = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday']

print('Dear Patient, Please follow below prescription') # greet patient and say it as description
for each_day in a_week:                                 # a loop to repeat and execute each day in a week
    print (each_day+', Eat 1 Apple...')                 # print the prescription, combine two strings with '+'

### Hello World

Lets start our Basic with a usual Hello World program in Python

In [ ]:
print("Hello Python World!!!")

In [ ]:
# let try some more print
print("I am a new programmer!!!")
print("Python is cool!!!")

Please note that either single quotes or double quotes shall be used in Print function. Both will give the desired output. But as a coding practise, lets follow using double quotes inside Print function

In [ ]:
print ('I an a new programmer!!!')

```python
# This line won't get executed because we prefixed it with a #

"""
Any Longer comments shall be written over multiple lines 
by enclosing them within triple
single or double quotes. Don't need to specify each line with #
"""


```

While learning any Programming language, main thing to understand is how to call for help, if stuck with issues.

In Python and in many GUI including our Jupyter Notebook, help can be attained in 2 ways. 

1. First, is to use the prefix or suffix `?`. 

2. Second, is to use `help()` to access the Help documentation. for instance if you need help on what Print function is, then use `help(print)`.


In [ ]:
#########################################################
################## Class Exercise 03_1 ##################
#########################################################

# Try using the ? to check the outcome on the print function. Just execute and see.

print?

In [ ]:
#########################################################
################## Class Exercise 03_2 ##################
#########################################################

# Try calling out help and find out the outcome. Just execute and see and below is sytax for using help.

help(print)

## Python as Calculator

Python is a Calculator. What the crap?. Python is way beyond our thinking.
Neverthless, let's start Python humbly

In [ ]:
1+2

In [ ]:
((2+3)/5) * ((3+6)/3)

In [ ]:
(1+2.5) * 5 / (2.75-0.25)

In [ ]:
2.25 ** 3   # ** - 2.25 to the power of 3 

In [ ]:
5+8
7.2+9.6
2.3
3+7

Only the result of last line will be displayed in the Out value and the other previous line values won't be displayed. To display the same, you can use the function <code>print</code>.

Now revise the query as below and try to see all the lines output

In [ ]:
print(5+8)
print(7.2+9.6)
print(2.3)
print(3+7)

In [ ]:
#########################################################
################## Class Exercise 03_3 ##################
#########################################################

<code>Question:</code> 
A dairy store had 1000 packs of milk in the storage for the day. Out of the total storage, 5% were spoiled, and around 500 milk packs were sold so far. Calculate how many packs of milk left in the dairy store for the day?

In [1]:
1000 - (1000*(5/100)) - 500

450.0

In [ ]:
1000*(1-0.05) - 500

# Variables, Objects and Data Types

Lets take the same class exercise as above and make it more clear and understanding to everyone by assigning each items to a variable.

<code>Question:</code> 
A dairy store had 1000 packs of milk in the storage for the day. Out of the total storage, 5% were spoiled, and around 500 milk packs were sold so far. Calculate how many packs of milk left in the dairy store for the day.

In [2]:
stock = 1000 # initialising a variable stock with value as 1000. Total milk pack for the day available in the store
spoiled = 0.05 # 5% 
Total_sold = 500

# to the calculation part
stock*(1-spoiled)-Total_sold

450.0

<div class="alert alert-block alert-warning">
<b>Note:</b> 
Please note that in Python, as like most of the languages, the "=" symbol is used as assignment operator. Always keep in mind that the equal '=' sign used while assigning a value to a variable shouldn't be read as 'equal to', It should be read or interpreted as "is set to". The variable is assigned with some value. In the above case, "Stock" variable is set to 1000 or assigned with 1000.
</div>

#### Variables
- container to store value.
- once assigned with value, can be reused.
- rather having only numbers, variables improves the readability of the code

#### Python variable naming conventions
<div class="alert alert-block alert-warning">
<b>Notes:</b> 
variable names should follow the below:

- Only one word, no space separated words allowed
- Only consist of letters, numbers, and underscores. Cannot contain Special characters
- Usage of numbers allowed in Variables, but the variables cannot begin with a number
- Avoid using with Python <a href="https://www.programiz.com/python-programming/keyword-list"><b>keywords</b></a> in variable names


Don't worry about memorizing all keywords. In Jupyter notebook or any other programming environments (IDE), the keywords are highlighted by a specific colors than normal words, which will prevent users from naming their variables and functions after the keywords. 
</div>

As per previous example, the initial stock was 1000, post sale and spoiled ones, the lates stock available was 450. But if you query, the stock variable, it will still show as 1000 which is the initial stock.

In [3]:
stock

1000

To update the new stock value post sale and spoiled ones, what ever calculation we performed earlier to figure out the latest stock should get assigned to the same stock variable.

In [4]:
stock=stock*(1-spoiled)-Total_sold

In [5]:
stock

450.0

The Assignment operator is not as simple as just "=" (equal to), it can even more. Like the below table.

Operators |   Example  |  Comments
:--------|:--------|:---------
<code>+=</code> | x+=y | The example is equal to x=x+y
<code>-=</code> | x-=y | The example is equal to x=x-y
<code>*=</code> | x*=y | The example is equal to x=x*y
<code>/=</code> | x/=y | The example is equal to x=x/y

For example, to showcase on other assignment operator, lets consider the same example. Currently, the stock is left with 450. Now, some time back, additional 50 milk packs got sold. so we need to update the latest stock left. We can perform below action as simple as like this.

In [6]:
stock -= 50 # this assigment operator is equal to "stock = stock - 50"

In [7]:
stock

400.0

## Objects

Python is an Object Oriented Programming Language. Everything in Python is Object.

In [8]:
7 # is an object in Python

7

Python Objects attributes includes identity, type and value

### Object ID

In [9]:
print (id(7)) # 7 is an object and it's identity is verified with id(). This is the address in Memory.

140354833803760


Don't get confused, just treat this as like your house address to identify where you living.
Even the built-in functions had object id

In [10]:
id(print)

140354565284784

### Object Data Type

In [11]:
type(7)

int

We will discuss in detail below on the Data Types

# Python can Speak

Lets see how Python can interact with us. Even similar way you can use the coding practice enjoyable.

In [12]:
# Python can speak too!, via coding.
print("Hello, I am Python! Let's have a conversation.")

name = input("What's your name? ")
print("Hello ",name,"!", "I'ts great to meet you.")

age = int(input("How old are you? "))
print("Wow, you are ", age, "years old! That's fantastic!")

feeling = input("How are you feeling today? ")
print("I understand. It's essential to express our emotions.")

hobby = input("What are your hobbies? ")
print(hobby,"sounds like an exciting way to spend your time!")

print("Thank you for chatting with me. Remember, Python is not only a programming language but also a great conversationalist!")

Hello, I am Python! Let's have a conversation.
What's your name? Learn Tech & Finance
Hello  Learn Tech & Finance ! I'ts great to meet you.
How old are you? 25
Wow, you are  25 years old! That's fantastic!
How are you feeling today? Awesome
I understand. It's essential to express our emotions.
What are your hobbies? Coding...
Coding... sounds like an exciting way to spend your time!
Thank you for chatting with me. Remember, Python is not only a programming language but also a great conversationalist!


In [13]:
help(input)

Help on method raw_input in module ipykernel.kernelbase:

raw_input(prompt='') method of ipykernel.ipkernel.IPythonKernel instance
    Forward raw_input to frontends
    
    Raises
    ------
    StdinNotImplementedError if active frontend doesn't support stdin.



###  Detailed documentation shared by Python.org
- <a href="https://docs.python.org/3/library/functions.html#input"><b>Python Documentation on Built-in Functions</b></a>

In Python, we use the input() function to take input from the user. Whatever you enter as input, the input function converts it into a string. If you enter an integer value still input() function converts it into a string.

In [ ]:
# The code is Python, but it can speak too!
print("Hello, I am Python! Let's have a conversation.")

name = input("What's your name? ")
print(f"Hello, {name}! It's great to meet you.")

age = int(input("How old are you? "))
print(f"Wow, you are {age} years old! That's fantastic!")

feeling = input("How are you feeling today? ")
print(f"I understand. It's essential to express our emotions.")

hobby = input("What are your hobbies? ")
print(f"{hobby} sounds like an exciting way to spend your time!")

print("Thank you for chatting with me. Remember, Python is not only a programming language but also a great conversationalist!")


### Other way around of using Print via format is below for reference.

In [14]:
print("My name is {0} and my age is {1} ".format(name,age))

My name is Learn Tech & Finance and my age is 25 


In [15]:
print("My name is {0} and my age is {1} ".format(age,name))

My name is 25 and my age is Learn Tech & Finance 


### Python Dynamicity

So far we have seen the variable which acts a storage and if you have some programming background, you could have noticed that there is no explicit declaration of variables as like in C, C++, Java.

``` python
### C Code Snipet

#include <stdio.h>

int main() {
    int age; // Variable declaration
    age = 25; // Variable assignment
    printf("Age: %d\n", age);
    return 0;
}

```

``` python
### C++ Code Snippet

#include <iostream>

int main() {
    int age; // Variable declaration
    age = 25; // Variable assignment
    std::cout << "Age: " << age << std::endl;
    return 0;
}

```

``` python
### Java Code Snippet

public class VariableDeclaration {
    public static void main(String[] args) {
        int age; // Variable declaration
        age = 25; // Variable assignment
        System.out.println("Age: " + age);
    }
}
```

Since Python is dynamic, you don't need to explicitly declare a variable and python have the capability to detect what kind of data type it is and accordingly it will take it up.
For the same example as like other languages (C, C++ and JAVA), Python's code snippet is below for reference.

In [16]:
age = 25 # Both Variable Declaration and Variable Assignment. No need for any explicit declaration
age

25

This will lead us to the next topic of <b> DATA TYPES </b>

## DATA TYPES

We haven't declared the data type for the above example <b> age </b> right. Let's see what's the data type python assigned to it

In [17]:
type (age)

int

Let's get into detail on the Python Data Types in the vido and then comeback to the notepad.

As we understood, Python has below Data Types for your kind reference over here itself.

``` python 
Text Type: str (a='Hi')
Numeric Type: int (a=5), float (a=5.0), complex (a=2k)
Boolean Type: bool (a=True, x=False)
Set Type: set (a={"dog","cat","fish"})
Mapping Type: dict (a={"name":"Alex","age":23})
Sequence Type: list (a=["dog","cat","fish"]), tuple (a=("dog","cat","fish")), range (a=range(5))

These are all basic built-in Data Types exists in Python
    
```

Some of the Data Types we will go through during our upcoming sessions. Lets go through some basics now.

### Numeric Data Type

Lets take our old dairy store example and see through the data types

<div class="alert alert-block alert-success">
<b>Question:</b> 
A dairy store had 1000 packs of milk in the storage for the day. Out of the total storage, 5% were spoiled, and around 500 milk packs were sold so far. Calculate how many packs of milk left in the dairy store for the day.
</div>

In [18]:
stock = 1000
type(stock)

int

In [19]:
spoiled = 0.05 # 5%
type(spoiled)

float

In [20]:
Total_sold = 500
type(Total_sold)

int

In [21]:
type(stock*(1-spoiled)-Total_sold)

float

In [22]:
complex_number = 3+4j

In [23]:
type(complex_number)

complex

### String Data Type

In [24]:
x = 'Hi '
y = 'Welcome '
print(x)
print(y)

Hi 
Welcome 


In [25]:
type(x)

str

In [26]:
type(y)

str

Anything enclosed between single or double quotes are considered as STRINGS in Python

In [27]:
z=x+y

In [28]:
z

'Hi Welcome '

In [29]:
type(z)

str

The "+" operator shall be used to combine to strings and at the same time "*" operator shall be used to repeat the string by the specified number of times

In [30]:
print(z*10)

Hi Welcome Hi Welcome Hi Welcome Hi Welcome Hi Welcome Hi Welcome Hi Welcome Hi Welcome Hi Welcome Hi Welcome 


In [31]:
print('Hi Everyone, Welcome \n' *10)

Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 
Hi Everyone, Welcome 



<div class="alert alert-block alert-warning">
<b>Note:</b> 
Please note that the multiplication (*) operator will work different for Numeric data type than String data types.
</div>

For example, you are going to take input from user and going to print that 10 times.

In [32]:
value = input('Type something: ')

print('Your input is: ' + (value)*10)

Type something: Hi
Your input is: HiHiHiHiHiHiHiHiHiHi


In [33]:
value = input('Type something: ')

print('Your input is: ' + (value)*10)

Type something: 10
Your input is: 10101010101010101010


In [34]:
type(value)

str

In [35]:
value = input('Type something: ')

print('Your input is: ' + str(float(value)*10))

Type something: 10
Your input is: 100.0


This is called data type conversion. That we will dicuss in the upcoming session

### Boolean

As like many other languages, Python Boolean data type denotes, "True" or "False". For now, understand only this much, we will revise everything in the upcoming session.

In [36]:
x = True

In [38]:
type(x)

bool

In [44]:
y = True

In [39]:
type(x)

bool

In [43]:
print(x==y)

False


Lets try with numbers and compare to see the output

In [46]:
x=2
y=3

In [47]:
print(x==y)

False


<div class="alert alert-block alert-warning">
<b>Note:</b> 
    Please note that the assignment operator is single "=" (equal to) and the comparator operator is double "=="
    (equal to)
</div>

### Data Type Conversion

You might be confused with What is "Data Type Conversion", on top of the loads taken so far in the basics of Python coding.

For instance, you have a requirement to append some string at the end of the integer data type variable.

In [48]:
n = 10
type(n)

int

In [49]:
o = n+'Hi'

TypeError: unsupported operand type(s) for +: 'int' and 'str'

To Address this issue, the integer data type variable 'n' shall be converted to a string first and then we can add string to the variable.

In [50]:
o=str(n)+'Hi'

In [51]:
o

'10Hi'

In [52]:
type(o)

str

Implicit conversion, Python will handle automatically and the example is below for reference.

In [53]:
x=2
y=3.5

In [54]:
type(x)

int

In [55]:
type(y)

float

In [56]:
z=x+y
z

5.5

In [57]:
type(z)

float

x is of integer and y is float, the summation of x and y will mark the variable z as 'float' automatically as it has decimal value.
Any other conversions, we need to specify explicitly as like previous example